File: src/ProductCatalogService/Models/ProductModel.cs
namespace ProductCatalogService.Models;

public class ProductModel
{
  public int Id { get; set; }
  public string ProductName { get; set; } = string.Empty;
  public string ProductDescription { get; set; } = string.Empty;
  public decimal Price { get; set; }
  public int Quantity { get; set; }
  public Status Status { get; set; } 
  public InventoryStatus InventoryStatus { get; set; }
  public Seo Seo { get; set; } = null!;
  public ICollection<Category> Categories { get; set; } = null!;
  public ICollection<Specification> Specifications { get; set; } = null!;
  public ICollection<Image> Images { get; set; } = null!;
  public ICollection<Variant> Variants { get; set; } = null!;
  public ICollection<string> RelatedProducts { get; set; } = null!;
  public ICollection<Review> Reviews { get; set; } = null!;
  public ICollection<Rating> Ratings { get; set; } = null!;
  public double AggregateRating
  {
    get
    {
      return Ratings.Any() ? Ratings.Average(r => r.RatingValue) : 0;
    }
  }

  public double ReviewCount
  {
    get
    {
      return Reviews.Count;
    }
  }
}

File: src/node-mongo/src/models/productModel.js
import mongoose from 'mongoose';
const { Schema } = mongoose;

const productSchema = new mongoose.Schema({
  id: { type: String },
  productName: { type: String, required: true },
  productDescription: { type: String, required: true },
  price: { type: Number, required: true },
  quantity: { type: Number, required: true },
  status: { type: String, default: 'In Stock' },
  inventoryStatus: String,
  seo: {
    metaTitle: String,
    metaDescription: String,
    slug: String,
  },
  categories: [{ type: Schema.Types.ObjectId, ref: 'Category' }],
  specifications: [
    {
      name: String,
      value: String,
    },
  ],
  images: [
    {
      imageUrl: String,
      imageDescription: String,
    },
  ],
  variants: [
    {
      id: String,
      color: String,
      size: String,
      quantity: Number,
    },
  ],
  relatedProducts: [String],
  reviews: [
    {
      customerId: String,
      comment: String,
    },
  ],
  ratings: [
    {
      customerId: String,
      rating: Number,
    },
  ],
  AggregateRating: Number,
  ReviewCount: Number,
});

// Define index for productName
productSchema.index({ productName: 'text' });

const Product = mongoose.model('Product', productSchema);

export default Product;
